File: php/src/Muflone/Persistence/ISerializer.php
<?php

declare(strict_types=1);

namespace Muflone\Persistence;

interface ISerializer
{
    /**
     * Deserialize data from a string
     *
     * @template T
     * @param string $serializedData
     * @param class-string<T> $className
     * @return T|null
     */
    public function deserialize(string $serializedData, string $className): ?object;

    /**
     * Serialize data to a string
     *
     * @param object $data
     * @return string
     */
    public function serialize(object $data): string;
}

File: src/Muflone/Persistence/ISerializer.cs
using System.Threading;
using System.Threading.Tasks;

namespace Muflone.Persistence;

public interface ISerializer
{
    Task<T?> DeserializeAsync<T>(string serializedData, CancellationToken cancellationToken = default) where T : class;
    Task<string> SerializeAsync<T>(T data, CancellationToken cancellationToken = default) where T : class;
}